### PR TITLE
[refactor] #3353: Remove eyre from block.rs

### LINKED ISSUE #3353: Remove eyre from iroha libraries
Using `eyre::Report` is only good for errors in binaries. For libraries we should have custom error types (that implement `thiserror::Error`)

### PR DESCRIPTION
**Pull request by Erigara**
## Description

Replace eyre error messages with structured errors.

### Linked issue

Partially closes #3353 

### Benefits

Less vague errors.

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [x] (optional) I've written unit tests for the code changes
- [x] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

**Review by mversic — APPROVED** [review]

File: core/src/block.rs [hunk 1]
```diff
@@ -12,15 +12,111 @@
...
+#[derive(Debug, Error)]
+pub enum TransactionRevalidationError {
+    /// Failed to accept transaction
+    #[error("Failed to accept transaction")]
+    Accept(#[from] AcceptTransactionFail),
+    /// Transaction isn't valid but must be
+    #[error("Transaction isn't valid but must be")]
+    NotValid(#[from] TransactionRejectionReason),
+    /// Rejected transaction in valid
+    #[error("Rejected transaction in valid")]
```
Comment (pesterev) [inline core/src/block.rs:117]:
```suggestion
    /// Rejected transaction is valid
    #[error("Rejected transaction is valid")]
```

File: core/src/block.rs [hunk 2]
```diff
@@ -640,6 +578,114 @@ impl Revalidate for VersionedCommittedBlock {
...
+                    .map_err(|(_tx, error)| error)
+                    .map_err(TransactionRevalidationError::NotValid)
+            })
+        })
+        .try_fold(Vec::new(), |mut acc, tx| {
+            tx.map(|valid_tx| {
+                acc.push(valid_tx);
+                acc
+            })
+        })?;
```
Comment (pesterev) [inline core/src/block.rs:647]:
```suggestion
        .collect::<Result<Vec<_>, _>>()?;
```

File: core/src/block.rs [hunk 3]
```diff
@@ -640,6 +578,114 @@ impl Revalidate for VersionedCommittedBlock {
...
+                    Ok(_) => Err(TransactionRevalidationError::RejectedIsValid),
+                },
+            )
+        })
+        .try_fold(Vec::new(), |mut acc, rejected_tx| {
+            rejected_tx.map(|tx| {
+                acc.push(tx);
+                acc
+            })
+        })?;
```
Comment (pesterev) [inline core/src/block.rs:684]:
```suggestion
        .collect::<Result<Vec<_>, _>>()?;
```

File: core/src/block.rs [hunk 4]
```diff
@@ -12,15 +12,111 @@
...
+    /// Block has committed transactions
+    #[error("Block has committed transactions")]
+    HasCommittedTransactions,
+    /// Mismatch between the actual and expected hashes of the latest block
+    #[error("Mismatch between the actual and expected hashes of the latest block. Expected: {:?}, actual: {:?}", expected, actual)]
+    LatestBlockHashMismatch {
+        /// Expected value
+        expected: Option<HashOf<VersionedCommittedBlock>>,
+        /// Actual value
+        actual: Option<HashOf<VersionedCommittedBlock>>,
```
Comment (0x009922) [inline core/src/block.rs:68]: Here and later with `expected` and `actual`, doc comments look redundant.

**Review by 0x009922 — APPROVED** [review]
LGTM. It will be easy to extend new error enums with `miette`s `Diagnostic`.

File: core/src/block.rs [hunk 5]
```diff
@@ -12,15 +12,111 @@
...
 
-use crate::{prelude::*, sumeragi::network_topology::Topology};
+/// Errors occurred on block revalidation
+#[derive(Debug, Error)]
+pub enum BlockRevalidationError {
+    /// Block is empty
+    #[error("Block is empty")]
+    Empty,
+    /// Block has committed transactions
+    #[error("Block has committed transactions")]
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply core/src/block.rs:60]: Created #3620

**Pull request merged**